### PR TITLE
rgw: fix rgw_data_sync_info::json_decode()

### PR DESCRIPTION
**Pull request by cbodley**
this was trying to decode a 'uint64_t instance_id' into the wrong variable 'uint32_t num_shards'. datalog trimming relies on this json format, and decode failures here prevent any trimming from taking place

Fixes: http://tracker.ceph.com/issues/38373

automated datalog trimming wasn't covered by regression testing, so i've added a 'radosgw-admin datalog autotrim' command to exercise the logic, and a simple multisite test case to verify it

### CODE REVIEW

File: src/rgw/rgw_data_sync.h [hunk 1]
```diff
@@ -86,7 +86,7 @@ struct rgw_data_sync_info {
       state = StateInit;
     }
     JSONDecoder::decode_json("num_shards", num_shards, obj);
-    JSONDecoder::decode_json("instance_id", num_shards, obj);
```
Comment (mattbenjamin) [inline src/rgw/rgw_data_sync.h]: oi vey

**Comment by cbodley** [comment]
> is datalog autotrim useful only for testing? doc is sparse

that's the intent. i hadn't documented the other mdlog/bilog autotrim commands, though i have found them useful when we need to capture a debug log of trim failures since it doesn't get mixed in with the rest of the radosgw/multisite log output

**Comment by mattbenjamin** [comment]
@cbodley ok, cool

**Comment by cbodley** [comment]
added an extra unrelated fix, because the `datalog autotrim` test was failing due to races with background data sync activity from other 'bucket sync disable' tests that didn't wait for data sync to complete

**Comment by cbodley** [comment]
http://pulpito.ceph.com/cbodley-2019-02-19_19:15:26-rgw-wip-rgw-datalog-autotrim-distro-basic-smithi/

all `test_datalog_autotrim` and enable/disable tests passed. a single multisite test failure in `test_versioned_object_incremental_sync`, and some unrelated valgrind leaks in `RGWSendRawRESTResourceCR` (possibly from async notification threads?)

**Pull request merged**